### PR TITLE
Added a simple instancing test; scene_graph makes the batches, render…

### PR DESCRIPTION
**Pull request by Nielsbishere**
…_system renders them. Updated shader to support instances and create render target now defaults to render target (not back buffer).

### CODE REVIEW

File: src/d3d12/d3d12_renderer.cpp [hunk 1]
```diff
@@ -180,7 +176,7 @@ namespace wr
 		d3d12::Signal(m_fences[frame_idx], m_direct_queue);
 	}
 
-	void D3D12RenderSystem::RenderSceneGraph(SceneGraph const & scene_graph)
+	void D3D12RenderSystem::RenderSceneGraph(SceneGraph& scene_graph)
```
Comment (VZout) [inline src/d3d12/d3d12_renderer.cpp]: This should really be const. You wouldn't expect changes in a render func.

File: src/d3d12/d3d12_renderer.cpp [hunk 2]
```diff
@@ -246,17 +234,52 @@ namespace wr
...
-		{
-			auto n_mesh = static_cast<D3D12Mesh*>(mesh);
-			auto frame_idx = GetFrameIdx();
+	}
+
+	//Render batches
+	void D3D12RenderSystem::RenderMeshBatches(SceneGraph& scene_graph)
+	{
+		
+		auto &batches = scene_graph.GetBatches();
```
Comment (VZout) [inline src/d3d12/d3d12_renderer.cpp]: I don't believe the style guide mentions this but please put the reference or pointer symbols to the left or centered. Everywhere else its either left or centered.

File: src/d3d12/d3d12_renderer.cpp [hunk 3]
```diff
@@ -246,17 +234,52 @@ namespace wr
...
+			scene_graph.Optimize();
+
+		//Render batches
+		for (auto &elem : scene_graph.GetBatches()) {
+
+			Model *model = elem.first;
+			temp::MeshBatch &batch = elem.second;
+
+			//Update object data
+			d3d12::UpdateConstantBuffer(batch.batchBuffer->m_native, GetFrameIdx(), &batch.data, sizeof(batch.data));
```
Comment (VZout) [inline src/d3d12/d3d12_renderer.cpp]: Don't update constant buffers here. Should preferably happen somewhere that makes more sense (a place/time where also other buffers get updated).

File: src/d3d12/d3d12_renderer.hpp [hunk 4]
```diff
@@ -3,16 +3,34 @@
 #include "../renderer.hpp"
 
 #include <DirectXMath.h>
+#include <unordered_map>
```
Comment (VZout) [inline src/d3d12/d3d12_renderer.hpp]: Unused?

File: src/d3d12/d3d12_renderer.hpp [hunk 5]
```diff
@@ -3,16 +3,34 @@
...
 
 #include "../vertex.hpp"
 #include "d3d12_structs.hpp"
 
 namespace wr
 {
-
+	struct Model;
```
Comment (VZout) [inline src/d3d12/d3d12_renderer.hpp]: Unused?

File: src/scene_graph/mesh_node.hpp [hunk 6]
```diff
@@ -1,10 +1,11 @@
 #pragma once
 
 #include "scene_graph.hpp"
+#include <unordered_map>
```
Comment (VZout) [inline src/scene_graph/mesh_node.hpp]: Unused?

File: src/scene_graph/scene_graph.cpp [hunk 7]
```diff
@@ -52,4 +59,32 @@ namespace wr
 		return nullptr;
 	}
 
+	temp::MeshBatches& SceneGraph::GetBatches() { return m_batches; }
+
+	void SceneGraph::Optimize() {
+
```
Comment (VZout) [inline src/scene_graph/scene_graph.cpp]: This function contains some new lines at the beginning and end.

File: src/scene_graph/scene_graph.hpp [hunk 8]
```diff
@@ -8,6 +8,8 @@
 #include "../util/defines.hpp"
 #include "../resource_pool.hpp"
 
+#define MAX_INSTANCE_COUNT 768
```
Comment (VZout) [inline src/scene_graph/scene_graph.hpp]: Prefer a `static const constexpr` in `d3d12_settings.hpp` over a definition

File: src/scene_graph/scene_graph.hpp [hunk 9]
```diff
@@ -59,12 +87,18 @@ namespace wr
 		void RemoveChildren(std::shared_ptr<Node> const & parent);
 		std::shared_ptr<CameraNode> GetActiveCamera();
 
+		void Optimize();
+		temp::MeshBatches &GetBatches();
```
Comment (VZout) [inline src/scene_graph/scene_graph.hpp]: Getter should be const.

File: src/scene_graph/scene_graph.hpp [hunk 10]
```diff
@@ -40,6 +42,32 @@ namespace wr
...
+		};
+
+		struct MeshBatch
+		{
+			unsigned int num_instances = 0;
+			D3D12ConstantBufferHandle* batchBuffer;
+			Model_CBData data;
+		};
+
+		typedef std::unordered_map<Model*, MeshBatch> MeshBatches;
```
Comment (VZout) [inline src/scene_graph/scene_graph.hpp]: Prefer `using` over `typedef`

File: src/d3d12/d3d12_renderer.cpp (hunk 1 above)
Comment (Nielsbishere) [reply src/d3d12/d3d12_renderer.cpp]: Not possible; SceneGraph::Optimize sets up the batches inside of RenderSceneGraph -> RenderMeshBatches; which requires write access.

File: src/d3d12/d3d12_renderer.cpp (hunk 3 above)
Comment (Nielsbishere) [reply src/d3d12/d3d12_renderer.cpp]: Moved to optimize

File: src/scene_graph/scene_graph.hpp (hunk 8 above)
Comment (Nielsbishere) [reply src/scene_graph/scene_graph.hpp]: That would require scene_graph.hpp to include d3d12_settings.hpp though

**Pull request closed without merging**